honeybadger-java/src/main/java/io/honeybadger/loader/HoneybadgerNoticeLoader.java

```diff
@@ -0,0 +1,121 @@
+package io.honeybadger.loader;
+
+import com.fasterxml.jackson.annotation.JsonInclude;
+import com.fasterxml.jackson.databind.InjectableValues;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.SerializationFeature;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import io.honeybadger.reporter.config.ConfigContext;
+import io.honeybadger.reporter.dto.Notice;
+import org.apache.http.HttpResponse;
+import org.apache.http.HttpStatus;
+import org.apache.http.client.fluent.Request;
+import org.apache.http.client.fluent.Response;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.net.URI;
+import java.util.UUID;
+
+/**
+ * Utility class used to load a fault's details into a readable object
+ * structure.
+ *
+ * This class is currently only suitable for consumption by integration tests,
+ * as much of the data in the reporter dto is not returned by the Notice API.
+ * If you have an use case for Notice deserialization, please file an issue
+ * at https://github.com/honeybadger-io/honeybadger-java
+ *
+ * @author <a href="https://github.com/dekobon">Elijah Zupancic</a>
+ * @since 1.0.9
+ */
+public class HoneybadgerNoticeLoader {
+    private static final int RETRIES = 3;
+    public static final int RETRY_DELAY_MILLIS = 5000;
+    private final Logger logger = LoggerFactory.getLogger(getClass());
+
+    private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper()
+                    .setSerializationInclusion(JsonInclude.Include.NON_NULL)
+                    .configure(SerializationFeature.WRITE_NULL_MAP_VALUES, false);
+
+    private ConfigContext config;
+
+    public HoneybadgerNoticeLoader(final ConfigContext config) {
+        this.config = config;
+    }
+
+    String pullFaultJson(final UUID faultId) throws IOException {
```

Comment:
I would recommend the following improvement to `HoneybadgerNoticeLoader`.

This will allow you to parse the incoming JSON data as a stream rather than parsing it to a String and then to a JSON data structure. Additionally, it uses `readValue(jsonText, ObjectNode.class)` instead of `readText(jsonText)` because this way the parser should throw an error if it isn't getting a proper JSON object (I don't feel strongly about this change - take it or reject as you see fit).

```diff
--- a/honeybadger-java/src/main/java/io/honeybadger/loader/HoneybadgerNoticeLoader.java
+++ b/honeybadger-java/src/main/java/io/honeybadger/loader/HoneybadgerNoticeLoader.java
@@ -15,8 +15,8 @@ import org.apache.http.client.fluent.Response;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
-import java.io.ByteArrayOutputStream;
 import java.io.IOException;
+import java.io.InputStream;
 import java.net.URI;
 import java.util.UUID;
 
@@ -47,7 +47,7 @@ public class HoneybadgerNoticeLoader {
         this.config = config;
     }
 
-    String pullFaultJson(final UUID faultId) throws IOException {
+    InputStream pullFaultJson(final UUID faultId) throws IOException {
         String readApiKey = config.getHoneybadgerReadApiKey();
 
         if (readApiKey == null) {
@@ -94,18 +94,18 @@ public class HoneybadgerNoticeLoader {
             }
         }
 
-        ByteArrayOutputStream out = new ByteArrayOutputStream();
-        httpResponse.getEntity().writeTo(out);
-
-        return out.toString();
+        return httpResponse.getEntity().getContent();
     }
 
     public Notice findErrorDetails(final UUID faultId) throws IOException {
-        String jsonText = pullFaultJson(faultId);
+        final ObjectNode originalJson;
+
+        try (InputStream jsonText = pullFaultJson(faultId)) {
+            originalJson = OBJECT_MAPPER.readValue(jsonText, ObjectNode.class);
+        }
 
         // HACK: Since our API is not symmetric, we do this in order to rename fields
         // and get *some* of the data that we sent.
-        JsonNode originalJson = OBJECT_MAPPER.readTree(jsonText);
         JsonNode cgiData = originalJson.get("web_environment");
         ((ObjectNode)originalJson.get("request"))
                 .replace("cgi_data", cgiData);
```

Reply:
As HoneyBadgerNoticeLoader is currently only used by tests, I have mixed feelings about this. Probably makes more sense to optimize the submitError code in HoneybadgerReporter (or sendToHoneybadger maybe).

Reply:
That makes sense. If you want the context of the Json that errored, you may want to embed the actual Json in the exception. I forget if Jackson now does this for you or not, but I seem to remember that it only shows you a subset of the overall JSON when there is a parsing error.